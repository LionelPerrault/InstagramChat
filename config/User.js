// HANDY METHODS FOR USER ROUTES

const db = require('./db'),
  bcrypt = require('bcrypt-nodejs'),
  { deletePost } = require('./Post'),
  { deleteGroup } = require('./Group'),
  { deleteCon } = require('./Message'),
  root = process.cwd(),
  { promisify } = require('util'),
  { rmdir } = require('fs'),
  { DeleteAllOfFolder } = require('handy-image-processor'),
  { intersectionBy } = require('lodash'),
  catchify = require('catchify')

/**
 * Returns ID of a user
 * @param {String} username Username
 */

const getId = async username => {
  let s = await db.query('SELECT id FROM users WHERE username=? LIMIT 1', [
    username,
  ])
  return s ? s[0].id : null
}

/**
 * Returns [what] of user ID
 *
 * eq. getWhat('username', id) => id's username
 *
 * @param {String} what Eq. Username
 * @param {String} id ID to be used to return [what]
 */
const getWhat = async (what, id) => {
  let s = await db.query(`SELECT ${what} FROM users WHERE id=? LIMIT 1`, [id])
  return s[0][what]
}

/**
 * creates a new user
 * @param {Object} User User details
 */
const create_user = async user => {
  let hash = bcrypt.hashSync(user.password)
  user.password = hash
  let [e, s] = await catchify(db.query('INSERT INTO users SET ?', user))
  e ? console.log(e) : null
  return s
}

/** changes password */
const change_password = async ({ password, id }) => {
  let hash = bcrypt.hashSync(password)
  let [e] = await catchify(
    db.query('UPDATE users SET password=? WHERE id=?', [hash, id])
  )
  return e ? false : true
}

/**
 * compares password
 * @param {String} password Password
 * @param {String} hash Hash to be compared with password
 * @returns {Boolean} Boolean
 */
const comparePassword = (password, hash) => {
  let comp = bcrypt.compareSync(password, hash)
  return comp
}

/**
 * Returns whether session is following user
 * @param {Number} session Session ID
 * @param {Number} user User
 */
const isFollowing = async (session, user) => {
  let is = await db.query(
    'SELECT COUNT(follow_id) AS is_following FROM follow_system WHERE follow_by=? AND follow_to=? LIMIT 1',
    [session, user]
  )
  return db.tf(is[0].is_following)
}

/**
 * Returns whether User is fav_by's favorite
 * @param {Number} fav_by Favorite By
 * @param {Number} user User ID
 */
const favouriteOrNot = async (fav_by, user) => {
  let s = await db.query(
    'SELECT COUNT(fav_id) AS fav_count FROM favourites WHERE fav_by=? AND user=?',
    [fav_by, user]
  )
  return db.tf(s[0].fav_count)
}

/**
 * Returns whether User is blocked by user
 * @param {Number} block_by Block By
 * @param {Number} user User ID
 */
const isBlocked = async (block_by, user) => {
  let s = await db.query(
    'SELECT COUNT(block_id) AS block_count FROM blocks WHERE block_by=? AND user=?',
    [block_by, user]
  )
  return db.tf(s[0].block_count)
}

/**
 * Deactivates user
 * @param {user} user User to deactivate
 * @param {Object} req Express' Req object
 * @param {Object} res Express' Res Object
 */
const deactivate = async (user, req, res) => {
  let posts = await db.query('SELECT post_id FROM posts WHERE user=?', [user]),
    groups = await db.query('SELECT group_id FROM groups WHERE admin=?', [
      user,
    ]),
    cons = await db.query(
      'SELECT con_id FROM conversations WHERE user_one=? OR user_two=?',
      [user, user]
    ),
    dltDir = promisify(rmdir),
    QLusers = JSON.parse(req.cookies.users),
    filtered = QLusers.filter(u => u.id != user)

  // DELETE ALL POSTS
  posts.map(async p => {
    await deletePost({
      post: p.post_id,
      user,
      when: 'user',
    })
  })

  // DELETE ALL GROUPS
  groups.map(async g => {
    await deleteGroup(g.group_id)
  })
  await db.query('DELETE FROM group_members WHERE member=? OR added_by=?', [
    user,
    user,
  ])

  // DELETE ALL CONVERSATIONS
  cons.map(async c => {
    await deleteCon(c.con_id)
  })

  await db.query('DELETE FROM tags WHERE user=?', [user])
  await db.query('DELETE FROM favourites WHERE fav_by=? OR user=?', [
    user,
    user,
  ])
  await db.query('DELETE FROM follow_system WHERE follow_by=? OR follow_to=?', [
    user,
    user,
  ])
  await db.query(
    'DELETE FROM notifications WHERE notify_by=? OR notify_to=? OR user=?',
    [user, user, user]
  )
  await db.query('DELETE FROM profile_views WHERE view_by=? OR view_to=?', [
    user,
    user,
  ])
  await db.query(
    'DELETE FROM recommendations WHERE recommend_by=? OR recommend_to=? OR recommend_of=?',
    [user, user, user]
  )
  await db.query('DELETE FROM hashtags WHERE user=?', [user])

  DeleteAllOfFolder(`${root}/dist/users/${user}/`)
  await dltDir(`${root}/dist/users/${user}`)

  await db.query('DELETE FROM users WHERE id=?', [user])

  res.cookie('users', `${JSON.stringify(filtered)}`)
  req.session.reset()
}

/**
 * Returns mutual users
 * Returns mutual users of session & user
 * @param {Number} session Session ID
 * @param {Number} user User ID
 */
const mutualUsers = async (session, user) => {
  let myFollowings = await db.query(
      'SELECT follow_system.follow_id, follow_system.follow_to AS user, follow_system.follow_to_username AS username, users.firstname, users.surname FROM follow_system, users WHERE follow_system.follow_by=? AND follow_system.follow_to = users.id',
      [session]
    ),
    userFollowers = await db.query(
      'SELECT follow_system.follow_id, follow_system.follow_by AS user, follow_system.follow_by_username AS username, users.firstname, users.surname FROM follow_system, users WHERE follow_system.follow_to=? AND follow_system.follow_by = users.id',
      [user]
    ),
    mutuals = intersectionBy(myFollowings, userFollowers, 'user')

  return mutuals
}

/**
 * Mention users
 * @param {String} str Text which will be used to get mentioned users
 * @param {Number} session sessionID
 * @param {Number} post PostID
 * @param {String} when For fn to have knowledge when users were mentioned
 */
const mentionUsers = async (str, session, post, when) => {
  let users = str.match(/[^|\s]?@[\d\w]+/g)

  if (users) {
    for (let h of users) {
      let hash = h.slice(1)
      if (hash.substr(0, 1) !== '@') {
        let [{ userCount }] = await db.query(
          'SELECT COUNT(id) AS userCount FROM users WHERE username=?',
          [hash]
        )
        let id = await getId(hash)

        if (userCount == 1 && id != session) {
          await db.query('INSERT INTO notifications SET ?', {
            notify_by: session,
            notify_to: id,
            post_id: post,
            type: when == 'post' ? 'mention_post' : 'mention_comment',
            notify_time: new Date().getTime(),
          })
        }
      }
    }
  }
}

module.exports = {
  getId,
  getWhat,
  create_user,
  change_password,
  comparePassword,
  isFollowing,
  favouriteOrNot,
  isBlocked,
  deactivate,
  mutualUsers,
  mentionUsers,
}
