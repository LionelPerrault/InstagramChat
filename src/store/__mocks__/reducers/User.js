export default {
  session: {
    id: 24,
    username: 'takkar',
  },
  user_details: {
    id: 24,
    username: 'takkar',
    firstname: 'iam_',
    surname: 'takkar',
    email: 'takkar@gmail.com',
    bio: '',
    account_type: 'public',
    email_verified: 'yes',
    joined: '1480114098767',
    twitter: '',
    facebook: '',
    github: 'https://github.com/yTakkar/',
    instagram: 'https://www.instagram.com/_faiyaz_shaikh',
    phone: '',
    website: '',
    isOnline: false,
    lastOnline: '',
  },
  tags: [
    {
      tag: 'traveler',
      user: 24,
    },
    {
      tag: 'nature-lover',
      user: 24,
    },
  ],
  mutualUsers: [
    {
      user: 11,
      username: 'nobita',
      firstname: 'nobita',
      surname: 'nobi',
      follow_id: 261,
      mutualUsersCount: 0,
    },
  ],
}
