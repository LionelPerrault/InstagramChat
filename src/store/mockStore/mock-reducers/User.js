export default {
  session: {
    id: 24,
    username: 'takkar',
  },
  user_details: {
    id: 7,
    username: 'ghalib',
    firstname: 'Mirza',
    surname: 'Ghalib',
    email: 'ghalib@gmail.com',
    bio: '',
    account_type: 'private',
    email_verified: 'yes',
    joined: '1514718748562',
    twitter: '',
    facebook: '',
    github: '',
    instagram: '',
    phone: '',
    website: '',
    isOnline: false,
    lastOnline: '',
  }
}
