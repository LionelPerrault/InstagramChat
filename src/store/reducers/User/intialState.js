import { uData } from '../../../utils/utils'

export default {
  session: {
    id: Number(uData('session')),
    username: uData('username'),
  },
  user_details: {
    id: null,
    username: '',
    firstname: '',
    surname: '',
    email: '',
    bio: '',
    account_type: '',
    email_verified: '',
    joined: '',
    twitter: '',
    facebook: '',
    github: '',
    instagram: '',
    phone: '',
    website: '',
    isOnline: false,
    lastOnline: '',
  },
  tags: [],
  mutualUsers: [],
}
