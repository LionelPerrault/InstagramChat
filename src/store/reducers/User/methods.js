import { uniqBy } from 'lodash'

export const addTag = (tags, t) => {
  tags.unshift(t)
  return uniqBy(tags, 'tag')
}

export const deleteTag = (tags, value) => tags.filter(t => t.tag != value)
