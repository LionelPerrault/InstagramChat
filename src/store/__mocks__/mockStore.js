import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

// mock-reducers
import User from './reducers/User'
import Notification from './reducers/Notification'
import Group from './reducers/Group'
import Follow from './reducers/Follow'
import Explore from './reducers/Explore'
import Post from './reducers/Post'
import Hashtag from './reducers/Hashtag'
import Setting from './reducers/Setting'
import Message from './reducers/Message'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore,
  Post,
  Hashtag,
  Setting,
  Message,
}

const middlewares = [thunk]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
