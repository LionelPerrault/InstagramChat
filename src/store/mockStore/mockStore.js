import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import promise from 'redux-promise-middleware'

// mock-reducers
import User from './mock-reducers/User'
import Notification from './mock-reducers/Notification'
import Group from './mock-reducers/Group'
import Follow from './mock-reducers/Follow'
import Explore from './mock-reducers/Explore'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore
}

const middlewares = [
  promise(), thunk, logger
]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
