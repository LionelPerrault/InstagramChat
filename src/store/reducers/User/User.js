/* eslint indent:0 */
/* eslint no-unreachable:0 */

import InitialState from './intialState'
import * as methods from './methods'

export default (state = InitialState, action) => {
  let py = action.payload

  switch (action.type) {
    case 'GET_USER_DETAILS':
      return {
        ...state,
        user_details: py.details,
        tags: py.tags,
      }
      break

    case 'ADD_TAG':
      return { ...state, tags: methods.addTag(state.tags, py) }
      break

    case 'DELETE_TAG':
      return { ...state, tags: methods.deleteTag(state.tags, py) }
      break

    case 'GET_MUTUAL_USERS':
      return { ...state, mutualUsers: py }
      break
  }

  return state
}
