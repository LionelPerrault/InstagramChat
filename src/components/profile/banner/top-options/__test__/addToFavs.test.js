import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import mockStore from '../../../../../store/mockStore/mockStore'
import BannerAddToFavs, { PureBannerAddToFavs } from '../addToFavs'
import MockDataElement from '../../../../../utils/__test__/mock-dataElement'
import User from '../../../../../store/mockStore/mock-reducers/User'
import { shallow } from 'enzyme'

describe('BannerAddToFavs Component', () => {
  const mockFn = jest.fn()

  MockDataElement()

  const comp = (
    <Provider store={mockStore}>
      <BannerAddToFavs
        toggleOptions={mockFn}
      />
    </Provider>
  )

  it('should match snapshot with null', () => {
    const tree = create(comp).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should mock addToFavs action when option is clicked', () => {
    const wrapper = shallow(
      <PureBannerAddToFavs
        toggleOptions={mockFn}
        id={7}
      />
    )
    wrapper.find('.add_fav').simulate('click', {
      preventDefault() {}
    })
  })

  it('should match snapshot', () => {
    User.user_details.id = 7
    const tree = create(comp).toJSON()
    expect(tree).toMatchSnapshot()
  })

})
