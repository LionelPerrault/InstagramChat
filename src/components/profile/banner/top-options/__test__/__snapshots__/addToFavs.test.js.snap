// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BannerAddToFavs Component should match snapshot 1`] = `
<li>
  <a
    className="add_fav"
    href="#"
    onClick={[Function]}
  >
    Add to favourites
  </a>
</li>
`;

exports[`BannerAddToFavs Component should match snapshot with null 1`] = `null`;
